src/system/group.cr

```diff
@@ -0,0 +1,195 @@
+  def initialize(gid : Int)
+    group_struct = LibC.getgrgid(Group.int_to_gid(gid))
+    raise NotFound.new("Group with gid '#{gid}', was not found.") if group_struct.null?
+
+    group_struct = group_struct.value
+    @name = String.new(group_struct.gr_name)
+    @gid = group_struct.gr_gid
+    slice = group_struct.gr_mem.to_slice_null_terminated(LibC::NGROUPS_MAX)
+    @member_names = Array.new(slice.size) { |idx| String.new(slice[idx]) }
+  end
+
+  # Initializes a group by the given groupname.
+  def initialize(groupname : String)
+    group_struct = LibC.getgrnam(groupname.check_no_null_byte)
+    raise NotFound.new("Group with name '#{groupname}', was not found.") if group_struct.null?
+
+    group_struct = group_struct.value
+    @name = String.new(group_struct.gr_name)
+    @gid = group_struct.gr_gid
+    slice = group_struct.gr_mem.to_slice_null_terminated(LibC::NGROUPS_MAX)
+    @member_names = Array.new(slice.size) { |idx| String.new(slice[idx]) }
+  end
+
+  # Returns the groups name.
+  #
+  # ```
+  # System::Group.get(0).name # => "root"
+  # ```
+  getter name : String
+
+  # Returns the group ID.
+  #
+  # ```
+  # System::Group.get("root").gid # => 0
+  # ```
+  getter gid : UInt32
+
+  # Returns an array of the groups members names.
+  #
+  # ```
+  # System::Group.get("root").member_names
+  # ```
+  getter member_names : Array(String)
+
+  # Yields the groups members names.
+  #
+  # ```
+  # System::Group.get("root").member_names() { |name| puts names }
+  # ```
+  def each_member_name(&block : String ->) : Nil
+    @member_names.each { |member| yield(member) }
+  end
+
+  # Returns an array of the groups members.
+  #
+  # ```
+  # System::Group.get("root").members
+  # ```
+  def members : Array(User)
+    @member_names.map { |member| User.get(member) }
+  end
+
+  # Yields the groups members.
+  #
+  # ```
+  # System::Group.get("root").members() { |user| puts user.name }
+  # ```
+  def each_member(&block : User ->) : Nil
+    each_member_name { |member| yield(User.get(member)) }
+  end
+
+  # See `Object#hash(hasher)`
+  def hash(hasher)
+    @gid.hash(hasher)
+  end
+
+  # Returns `true` if `self` is equal to *other*.
+  def ==(other : Group) : Bool
+    @gid == other.gid
+  end
+
+  # Compares this group with *other*, returning `-1`, `0` or `+1` depending if the
+  # group ID is less, equal or greater than the *other* group ID.
+  def <=>(other : Group) : Int
+    @gid <=> other.gid
+  end
+
+  # Appends the username to the given `IO`.
+  def to_s(io : IO)
+    io << @name
+  end
+
+  # :nodoc:
+  def self.check_gid_in_bounds(gid : Int) : Nil
+    return if (gid >= 0 && gid <= Limits::GID_MAX)
+    raise OutOfBounds.new("gid: '#{gid}' is out of bounds.")
+  end
+
+  # :nodoc:
+  def self.int_to_gid(gid : Int) : LibC::GidT
+    check_gid_in_bounds(gid)
+    {% if flag?(:linux) || flag?(:darwin) || flag?(:freebsd) || flag?(:openbsd) %}
+      return gid.to_u32
+    {% else %}
+      return gid.to_u16
```

Comment:
ditto

Reply:
This isn't redundant. The standard is uint16_t many OS's change this to uint32_t. So if somebody adds a new OS it should be supported as the standard unless otherwise stated.

Reply:
Fixed despite my preference for explicit returns.